google/cloud/internal/rest_carrier.cc

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/internal/rest_carrier.h"
+#include "google/cloud/internal/rest_context.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+opentelemetry::nostd::string_view RestCarrier::Get(
+    opentelemetry::nostd::string_view) const noexcept {
+  // Since the client is never extracting data from the REST headers, we are not
+  // implementing Get and returning an empty string by default.
```

Comment:
Observation:  I guess that is a good thing, because it looks like `Get()` would be unimplementable.  This is also indicated by `Set()` being implemented by an "append to vector".  Is there some impedance mismatch between `TextMapCarrier` and `RestContext` that deserves closer attention?

Reply:
Correct observation. 

Interesting... there is a mismatch. A TextMapCarrier is expected to be a map<string, string> and HTTP headers are a map<string,vector<string>>.

I think since this should only be used to set the X-Cloud-Trace header, it is ok that it is "append to a vector" since that *should* only get set once, and therefore never append to the vector.

Alternatives options:
1) Create a SetHeader function for the REST Context that overwrites whatever value is in headers_ so that there is a single value for the key. 

2) Store a map of single value keys in the REST Context and multi-value keys and merge them on a call to headers()

@dbolduc, thoughts?

Reply:
> "We don't need to extract context in the client" 

For REST, note for Pub/Sub we will be using the Extract/Get.

The issue @devbww is pointing out is with the current implementation it expects us to only call `Set` once or else it might create headers that look like:

```
{"X-Cloud-Trace-Context" : {"context1", "context1", "context1", ....}
```

I think it makes sense to add something to avoid this. But I was wondering if there are other cases for single-value keys, because then we could store them separately in the headers and then we also could properly implement `Get`. But if not, it might be ok to do something more hack-y and just overwrite any values when set is called.

e.g. 

```
RestContext context; // headers =  {}
RestCarrier carrier(context);
carrier.Set("X-Cloud-Trace-Context", {"context1"}) // headers =  {"X-Cloud-Trace-Context" : {"context1"}}
carrier.Set("X-Cloud-Trace-Context", {"context2"}) // headers =  {"X-Cloud-Trace-Context" : {"context2"}}
``` 

instead of 

```
RestContext context; // headers =  {}
RestCarrier carrier(context);
carrier.Set("X-Cloud-Trace-Context", {"context1"}) // headers =  {"X-Cloud-Trace-Context" : {"context1"}}
carrier.Set("X-Cloud-Trace-Context", {"context1"}) // headers =  {"X-Cloud-Trace-Context" : {"context1", "context1}}
```

Reply:
> I think it makes sense to add something to avoid this.

I don't.

The propagators we care about do not set the same key twice. And we are not silly enough to call `TextMapPropagator::Inject()` multiple times in the same call.

Reply:
You convinced me with "we are not silly enough to call TextMapPropagator::Inject() multiple times in the same call."

I think the *key* is we don't control the `TextMapPropagator::Inject()` implementation. I'm modified it to overwrite whatever is in the header with a single value. 

Either way it's internal, so hopefully no one will call it twice.